datalad/plugin/addurls.py

```diff
@@ -0,0 +1,654 @@
+                                       "consider using {_repindex}"))
+        return
+
+    if dry_run:
+        for subpath in subpaths:
+            lgr.info("Would create a subdataset at %s", subpath)
+        for row in rows:
+            lgr.info("Would download %s to %s",
+                     row["url"], os.path.join(dataset.path, row["filename"]))
+            lgr.info("Metadata: %s", row["meta_args"])
+        yield get_status_dict(action="addurls",
+                              ds=dataset,
+                              status="ok",
+                              message="dry-run finished")
+        return
+
+    if not dataset.repo:
+        # Populate a new dataset with the URLs.
+        for r in dataset.create(result_xfm=None, return_type='generator'):
+            yield r
+
+    annex_options = ["--fast"] if fast else []
+
+    for spath in subpaths:
+        if os.path.exists(os.path.join(dataset.path, spath)):
+            lgr.warning(
+                "Not creating subdataset at existing path: %s",
+                spath)
+        else:
+            for r in dataset.create(spath, result_xfm=None,
+                                    return_type='generator'):
+                yield r
+
+    pbar_addurl = ui.get_progressbar(total=len(rows), label="Adding files",
+                                     unit=" Files")
+    pbar_addurl.start()
+
+    files_to_add = []
+    meta_to_add = []
+    addurl_results = []
+    for row_idx, row in enumerate(rows, 1):
+        pbar_addurl.update(row_idx)
+        fname_abs = os.path.join(dataset.path, row["filename"])
+        if row["subpath"]:
+            # Adjust the dataset and filename for an `addurl` call
+            # from within the subdataset that will actually contain
+            # the link.
+            ds_current = Dataset(os.path.join(dataset.path, row["subpath"]))
+            ds_filename = os.path.relpath(fname_abs, ds_current.path)
+        else:
+            ds_current = dataset
+            ds_filename = row["filename"]
+
+        if os.path.exists(fname_abs) or os.path.islink(fname_abs):
+            if ifexists == "skip":
+                addurl_results.append(
+                    get_status_dict(action="addurls",
+                                    ds=ds_current,
+                                    type="file",
+                                    path=fname_abs,
+                                    status="notneeded"))
+                continue
+            elif ifexists == "overwrite":
+                lgr.debug("Removing %s", fname_abs)
+                os.unlink(fname_abs)
+            else:
+                lgr.debug("File %s already exists", fname_abs)
+
+        try:
+            ds_current.repo.add_url_to_file(ds_filename, row["url"],
+                                            batch=True, options=annex_options)
+        except AnnexBatchCommandError:
+            addurl_results.append(
```

Comment:
so you aren't yielding because it would obscure the progress bar?
I have started to wait on an exciting results while running it on 10k files, but decided to check -- nothing was downloaded. apparently the issue is that tricky me didn't specify the filename -- only a directory with the hope that magic would happen and disposition filename or url filename would be used, but nope. So now it hides reports on errors:

```shell
$> datalad plugin addurls url_file=phenotype_mri_head10.csv url_format='{MRIs}' filename_format='{project}/sub-{participant_id}/ses-{session_id}_rep-{_repindex}/' fast=True    
                                                                                                                         [ERROR  ] failed to add URL [addurls(/tmp/testaddurl/abide_initiative/sub-50002/ses-1_rep-0/)] 
addurls(error): /tmp/testaddurl/abide_initiative/sub-50002/ses-1_rep-0/ (file) [failed to add URL]
[ERROR  ] failed to add URL [addurls(/tmp/testaddurl/abide_initiative/sub-50003/ses-1_rep-0/)] 
addurls(error): /tmp/testaddurl/abide_initiative/sub-50003/ses-1_rep-0/ (file) [failed to add URL]
...
action summary:                               
  addurls (error: 9)
```

if you want to maintain our super nice progressbar pristine, you could report # of failed within progressbar title, like we do during get:
```
$> datalad install -g -J20 ///labs/gobbini/famface/data
[INFO   ] Cloning http://datasets.datalad.org/labs/gobbini/famface/data to '/tmp/data' 
[INFO   ] access to dataset sibling "datasets.datalad.org" not auto-enabled, enable with:
| 		datalad siblings -d "/tmp/data" enable -s datasets.datalad.org 
install(ok): /tmp/data (dataset)
[INFO   ] Actually getting 543 files 
Total (1 ok, 61 failed out of 543):   0%|            ...
```
to alert the user that something is going wrong

Reply:
> so you aren't yielding because it would obscure the progress bar?

Yeah, I'm probably going about things wrong.  Will look more closely at how other places in datalad handle it.